### PR TITLE
[FrameworkBundle] Integrate the HtmlSanitizer component

### PR DESCRIPTION
**Pull request by tgalopin**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR adds the integration if the HtmlSanitizer component in the FrameworkBundle. See https://github.com/symfony/symfony/pull/44681 for details about the component.

The configuration for this integration is the following:

```yaml
framework:
    # This configuration is not required: as soon as you install the component, a default
    # "html_sanitizer" service is created, with the safe configuration, to be used directly.
    #
    # This configuration allows to set custom behaviors, in addition or instead of the default.
    html_sanitizer:
        # Default sanitizer (optional)
        # When not provided, the native "html_sanitizer" service is wired as default.
        default: my.sanitizer

        # Custom sanitizers (optional)
        sanitizers:

            # Each sanitizer defines its own service (no prefix/suffix) to ease understanding
            # Each sanitizer also defines a named autowiring alias to ease injection using variable name.

            # Here, this sanitizer is available as a service "my.sanitizer" or using autowiring
            # as "HtmlSanitizerInterface $mySanitizer".
            my.sanitizer:
                allow_safe_elements: true
                allow_elements:
                    iframe: ['src']
                    custom-tag: ['data-attr']
                    custom-tag-2: '*'
                block_elements:
                    - section
                drop_elements:
                    - video
                allow_attributes:
                    src: ['iframe']
                    data-attr: '*'
                drop_attributes:
                    data-attr: '*'
                force_attributes:
                    a:
                        rel: noopener noreferrer
                    h1:
                        class: bp4-heading
                force_https_urls: true
                allowed_link_schemes: ['http', 'https', 'mailto']
                allowed_link_hosts: ['symfony.com']
                allow_relative_links: true
                allowed_media_schemes: ['http', 'https', 'data']
                allowed_media_hosts: ['symfony.com']
                allow_relative_medias: true

            # "all.sanitizer" / "HtmlSanitizerInterface $allSanitizer"
            all.sanitizer:
                allow_all_static_elements: true
                allow_elements:
                    custom-tag: ['data-attr']

```

This PR is still WIP (esp tests) but I wanted to gather feedback regarding the configuration and DX as soon as possible.

### CODE REVIEW

**Comment by derrabus** [comment]
Remember to add `symfony/html-sanitizer` to the `require-dev` section of FrameworkBundle's composer.json.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                        ->scalarNode('default')
+                            ->defaultValue('html_sanitizer')
+                            ->info('Default sanitizer to use when injecting without named binding.')
+                        ->end()
+                        ->arrayNode('sanitizers')
+                            ->normalizeKeys(false)
+                            ->useAttributeAsKey('name')
+                            ->arrayPrototype()
+                                ->children()
+                                    ->scalarNode('allow_safe_elements')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: shouldn't this be a boolean node ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->info('Allows "safe" elements and attributes.')
+                                    ->end()
+                                    ->scalarNode('allow_all_static_elements')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: shouldn't this be a boolean node ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->info('Allows all static elements and attributes from the W3C Sanitizer API standard.')
+                                    ->end()
+                                    ->arrayNode('allow_elements')
+                                        ->info('Configures elements as allowed. Allowed elements are elements the sanitizer should retain from the input.')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: What is the expected structure of this array node ? Seeing it as a map with variable nodes as values looks weird to me if it is only about allowing elements.

If the values are a list of allowed attributes, this should be explained, and the config should use an array node with a scalar prototype instead of a variable node.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->normalizeKeys(false)
+                                        ->useAttributeAsKey('name')
+                                        ->defaultValue([])
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: this is useless. It is already the default value for a prototyped array node (same for all other prototyped nodes).

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->prototype('variable')
+                                        ->end()
+                                    ->end()
+                                    ->arrayNode('block_elements')
+                                        ->info('Configures elements as blocked. Blocked elements are elements the sanitizer should remove from the input, but retain their children.')
+                                        ->normalizeKeys(false)
+                                        ->useAttributeAsKey('name')
+                                        ->defaultValue([])
+                                        ->prototype('variable')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: What are the values in that array ? Shouldn't it be a list of scalar nodes instead of a map of variable nodes ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->end()
+                                    ->end()
+                                    ->booleanNode('force_https_urls')
+                                        ->info('Transforms URLs using the HTTP scheme to use the HTTPS scheme instead.')
+                                        ->defaultFalse()
+                                    ->end()
+                                    ->arrayNode('allowed_link_schemes')
+                                        ->info('Allows only a given list of schemes to be used in links href attributes.')
+                                        ->normalizeKeys(false)
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: should be removed as it is not a map-like node.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->defaultValue([])
+                                        ->prototype('variable')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: should be a scalar prototype

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                    ->end()
+                                    ->arrayNode('allow_elements')
+                                        ->info('Configures elements as allowed. Allowed elements are elements the sanitizer should retain from the input.')
+                                        ->normalizeKeys(false)
+                                        ->useAttributeAsKey('name')
+                                        ->defaultValue([])
+                                        ->prototype('variable')
+                                        ->end()
+                                    ->end()
+                                    ->arrayNode('block_elements')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:2164]: The sanitizer node is missing the `->fixXmlConfig` calls for all the prototyped children.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -2043,4 +2045,139 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
...
+                                        ->normalizeKeys(false)
+                                        ->defaultValue([])
+                                        ->prototype('variable')
+                                        ->end()
+                                    ->end()
+                                    ->arrayNode('allowed_link_hosts')
+                                        ->info('Allows only a given list of hosts to be used in links href attributes.')
+                                        ->normalizeKeys(false)
+                                        ->defaultValue([])
+                                        ->prototype('variable')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Same here than for allowed schemes.

Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: and same for other `allowed_*` nodes.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: you should add `->defaultFalse()`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 10]
```diff
@@ -2549,6 +2560,100 @@ private function registerUidConfiguration(array $config, ContainerBuilder $conta
...
 
+    private function registerHtmlSanitizerConfiguration(array $config, ContainerBuilder $container, PhpFileLoader $loader)
+    {
+        $loader->load('html_sanitizer.php');
+
+        foreach ($config['sanitizers'] as $sanitizerName => $sanitizerConfig) {
+            $def = $container->register($sanitizerName.'.config', HtmlSanitizerConfig::class);
+
+            // Base configuration
+            if ($sanitizerConfig['allow_safe_elements'] ?? false) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: this default value should be handled by the configuration class

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 11]
```diff
@@ -2549,6 +2560,100 @@ private function registerUidConfiguration(array $config, ContainerBuilder $conta
...
+            // Configures elements
+            foreach ($sanitizerConfig['allow_elements'] ?? [] as $element => $attributes) {
+                $def->addMethodCall('allowElement', [$element, $attributes], true);
+            }
+
+            foreach ($sanitizerConfig['block_elements'] ?? [] as $element) {
+                $def->addMethodCall('blockElement', [$element], true);
+            }
+
+            foreach ($sanitizerConfig['drop_elements'] ?? [] as $element) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: All those `?? []` can be removed. The configuration class already handles the default value.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 5 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Looking at the DI extension, it indeed expects a list of strings for `block_elements` and `drop_elements`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 12]
```diff
@@ -2549,6 +2560,100 @@ private function registerUidConfiguration(array $config, ContainerBuilder $conta
...
+
+            // Force attributes
+            foreach ($sanitizerConfig['force_attributes'] ?? [] as $element => $attributes) {
+                foreach ($attributes as $attrName => $attrValue) {
+                    $def->addMethodCall('forceAttribute', [$element, $attrName, $attrValue], true);
+                }
+            }
+
+            // Settings
+            if (isset($sanitizerConfig['force_https_urls'])) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: it will always be set thanks to the default value (same for others)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 13]
```diff
@@ -2549,6 +2560,100 @@ private function registerUidConfiguration(array $config, ContainerBuilder $conta
...
+            foreach ($sanitizerConfig['with_attribute_sanitizers'] ?? [] as $serviceName) {
+                $def->addMethodCall('withAttributeSanitizer', [new Reference($serviceName)], true);
+            }
+
+            foreach ($sanitizerConfig['without_attribute_sanitizers'] ?? [] as $serviceName) {
+                $def->addMethodCall('withoutAttributeSanitizer', [new Reference($serviceName)], true);
+            }
+
+            // Create the sanitizer and link its config
+            $container->register($sanitizerName, HtmlSanitizer::class)
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Not adding a prefix for the service seems weird to me. When we have named configuration in Symfony and in bundles, we generally don't require uniqueness over the whole container. So this makes it easy to override another service by mistake.

I suggest registering them as `html_sanitizer.sanitizer.<name>` instead, with `html_sanitizer.config.<name>` for the config.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 14]
```diff
@@ -2549,6 +2560,100 @@ private function registerUidConfiguration(array $config, ContainerBuilder $conta
...
+                ->addArgument(new Reference($sanitizerName.'.config'))
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Registering the config using a suffix on the user-defined name while the sanitizer service does not use any suffix makes FrameworkBundle fragile. If someone registers a sanitizer name ending in `.config`, it will override the config service of another sanitizer and break things.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 15]
```diff
@@ -802,4 +803,63 @@
...
+        <xsd:attribute name="force-https-urls" type="xsd:boolean" />
+        <xsd:attribute name="allow-relative-links" type="xsd:boolean" />
+        <xsd:attribute name="allow-relative-medias" type="xsd:boolean" />
+    </xsd:complexType>
+
+    <xsd:complexType name="allow-element">
+        <xsd:sequence>
+            <xsd:element name="attribute" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:attribute name="name" type="xsd:string" />
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: the attribute holding the map key should be marked as required in the XSD (because it is indeed required in the XML file, *before* merging files together)

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 16]
```diff
@@ -802,4 +803,63 @@
...
+
+    <xsd:complexType name="html-sanitizer">
+        <xsd:sequence>
+            <xsd:element name="sanitizer" type="sanitizer" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:attribute name="enabled" type="xsd:boolean" />
+        <xsd:attribute name="default" type="xsd:string" />
+    </xsd:complexType>
+
+    <xsd:complexType name="sanitizer">
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:832]: This type is missing the `name` attribute (which should be marked as required)

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 17]
```diff
@@ -802,4 +803,63 @@
...
+            <xsd:element name="block-element" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:sequence>
+            <xsd:element name="drop-element" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:sequence>
+            <xsd:element name="allow-attribute" type="allow-attribute" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:sequence>
+            <xsd:element name="drop-attribute" type="drop-attribute" minOccurs="0" maxOccurs="unbounded" />
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: undefined type

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 18]
```diff
@@ -802,4 +803,63 @@
...
+        </xsd:sequence>
+        <xsd:sequence>
+            <xsd:element name="force-attribute" type="force-attribute" minOccurs="0" maxOccurs="unbounded" />
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:839]: undefined type

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/full.xml [hunk 19]
```diff
@@ -39,5 +39,14 @@
             </framework:default-context>
         </framework:serializer>
         <framework:property-info />
+        <framework:html-sanitizer>
+            <framework:format name="csv">
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/full.xml]: invalid configuration

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php [hunk 20]
```diff
@@ -51,10 +51,10 @@ class Connection
     private bool $autoSetup;
     private int $maxEntries;
     private int $redeliverTimeout;
-    private int $nextClaim = 0;
-    private mixed $claimInterval;
-    private mixed $deleteAfterAck;
-    private mixed $deleteAfterReject;
+    private float $nextClaim = 0.0;
```
Comment (stof) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php]: Your PR seems to include unrelated changes

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 13 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: we do so on httpclient and it works well
we also do add a suffix for workflows, and it forces a needless convention on named autowiring aliases
I prefer not forcing any convention and showing examples to ppl, eg `my.sanitizer` as name

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: there's no way around such risks if we don't force a convention on the sanitizer name
but the risk is theoretical to me, nobody will use that conflicting name in practice

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 13 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I'm not asking to change the autowiring aliases here. Only the service names.

And if we require those names to avoid conflicts with any service in the container, we must document it explicitly in the info of the configuration, which is not the case right now.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 14 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Well, I know it happened in the past for another component (I don't remember which one as that was years ago), with someone reporting a very weird bug.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 13 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I reopened https://github.com/symfony/symfony/issues/41198#issuecomment-1002542381 so that we can have the discussion

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 3 above)
Comment (tgalopin) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I don't know how to use union types here: is there a possibility to require `array|string` as prototype?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 13 above)
Comment (tgalopin) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I fixed this by renaming services to:

```
html_sanitizer.sanitizer.my.sanitizer
html_sanitizer.config.my.sanitizer
```

And keeping the named autowiring alias to `$mySanitizer`

**Review by nicolas-grekas — COMMENTED** [review]
What's the status here?
I think we're missing a way to configure HtmlSanitizer's $maxInputLength.
The default seems low btw (20000). Shouldn't it be more like 1000000?

**Comment by fabpot** [comment]
@tgalopin Would be great (required?) if you can finish this PR before te 6.1 deadline which is happening soon now :)

**Comment by wouterj** [comment]
I've pushed a commit finalizing the tests and XML support. This PR should be ready to merge

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @tgalopin.
